### ISSUE TITLE
cytoolz.curried.sliding_window fails?

### ISSUE THREAD

**Issue by jni** [pytoolz/cytoolz#69]
```python
In [22]: from toolz import curried as C

In [23]: C.sliding_window(2)
Out[23]: <function sliding_window at 0x106673048>
```

vs

```python
In [19]: from cytoolz import curried as CC

In [20]: CC.sliding_window(2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-20-dfdf6311db18> in <module>()
----> 1 CC.sliding_window(2)

/Users/jni/conda/envs/anaconda/lib/python3.4/site-packages/cytoolz/functoolz.pyx in cytoolz.functoolz.curry.__call__ (cytoolz/functoolz.c:3300)()
    254             if required_args == -1 or len(args) < required_args:
    255                 return curry(self.func, *args, **kwargs)
--> 256         raise val
    257
    258     def __get__(self, instance, owner):

TypeError:
```

any ideas?

**Comment by eriknw** [comment]
Shoot.  Yeah, I have an idea what's going on.  Thanks for reporting this.  I'm surprised this issue has slipped through the cracks.

This is a "feature" of Python 3.4.  When inspecting a callable, it tries to report the number of required arguments even for builtin C extensions.  In other versions of Python, this inspection returns -1 if it can't determine the number.  Python 3.4 lies if it can't determine the number: it returns 0.  It even returns 0 for `map`, `filter`, etc.

I have explored an explicit way around this in #46 and pytoolz/toolz#202.  This made execution of `curry` *slightly* slower, but I think this may be worth it.  Two other options are to update the heuristics to `curry` to not raise the TypeError if the numargsrequired is 0, or to explicitly list all the curried cytoolz exceptions in the `known_numargs` workaround in `functoolz` for this issue.  I favor the explicit and generally applicable `numargs` keyword to `curry`.

**Comment by jni** [comment]
@eriknw Fascinating! Thank you for the detailed discussion!

Is the execution slower when creating the function, or every time a curried function is used?

Regarding the options, why would one ever curry a function with 0 required arguments? I see that numargs is useful generally, as you described, but I also think that not raising an error for 0 is also a good approach.

Last question: is it really a "feature"? Has this been raised as a possible Python bug?

**Comment by eriknw** [comment]
> Is the execution slower when creating the function, or every time a curried function is used?

When it's used, but the difference really is quite small.

> Regarding the options, why would one ever curry a function with 0 required arguments? I see that numargs is useful generally, as you described, but I also think that not raising an error for 0 is also a good approach.

Fair point.  If we can trust the 0 required arguments, then `curry` is able to re-raise actual `TypeError` exceptions.  This may occur in regular use when currying functions that only have keyword arguments.

> Last question: is it really a "feature"? Has this been raised as a possible Python bug?

It's an incomplete feature.  The expectation is that builtins will report their correct signature in future versions of Python.  I view defaulting to 0 as a bug.  IIRC, there have been bug reports about this, and the outlook was for it to be addressed in the future.

On a related note, I would expect builtins developed with Cython to eventually support the protocol that reports the correct signature.  I don't know the status on this.

**Comment by eriknw** [comment]
Fixed in #76.  Closing.

**Comment by jni** [comment]
Woot! Thank you! =)
